evalml/pipelines/component_graph.py

```diff
@@ -0,0 +1,335 @@
+        is only useful where the final component has only one parent, otherwise there will be
+        information missing.
+
+        Arguments:
+            X (pd.DataFrame): Data of shape [n_samples, n_features]
+
+        Returns:
+            pd.DataFrame: Transformed values.
+        """
+        return self._compute_features(self.compute_order[:-1], X)
+
+    def _compute_features(self, component_list, X, y=None, fit=False):
+        """Transforms the data by applying the given components.
+
+        Arguments:
+            component_list (list): The list of component names to compute.
+            X (pd.DataFrame): Input data to the pipeline to transform.
+            y (pd.Series): The target training data of length [n_samples]
+            fit (bool): Whether to fit the estimators as well as transform it.
+                        Defaults to False.
+
+        Returns:
+            pd.DataFrame or pd.Series - Output of the last component
+        """
+        output_cache = {}
+        final_component = component_list[-1]
+        for component_name in component_list:
+            component_class = self.get_component(component_name)
+            if not isinstance(component_class, ComponentBase):
+                raise ValueError('All components must be instantiated before fitting or predicting')
+            x_inputs = []
+            y_input = None
+            for parent_input in self.get_parents(component_name):
+                if parent_input[-2:] == '.y':
+                    if y_input is not None:
+                        raise ValueError(f'Cannot have multiple `y` parents for a single component {component_name}')
+                    y_input = output_cache[parent_input]
+                else:
+                    x_inputs.append(output_cache[parent_input])
+            input_x, input_y = self._merge(x_inputs, y_input, X, y)
+            if isinstance(component_class, Transformer):
+                if fit:
+                    output = component_class.fit_transform(input_x, input_y)
+                else:
+                    output = component_class.transform(input_x, input_y)
+                if isinstance(output, tuple):
+                    output_x, output_y = output[0], output[1]
+                else:
+                    output_x = output
+                    output_y = None
+                output_cache[f"{component_name}.x"] = output_x
+                output_cache[f"{component_name}.y"] = output_y
+            else:
+                if fit:
+                    component_class = component_class.fit(input_x, input_y)
+                output = component_class.predict(input_x)
+                output_cache[component_name] = output
+        final_component_class = self.get_component(final_component)
+        if isinstance(final_component_class, Transformer):
+            return output_cache[f"{final_component}.x"]
+        else:
+            return output_cache[final_component]
```

Comment:
Hm, I thought I mentioned this somewhere but it may actually have only been in my head - it doesn't make sense for an estimator to have either an "x" or "y" designation - if an estimator is not the last step in a pipeline, its "y" output would actually be used as an "x" input for another component or estimator, not as an actual target value. Because of that, estimators do not use the ".x" or ".y" suffixes at all.